presto-client/src/main/java/com/facebook/presto/client/SpnegoHandler.java

```diff
@@ -0,0 +1,330 @@
+        }
+        catch (ClientException ignored) {
+            return chain.proceed(chain.request());
+        }
+    }
+
+    @Override
+    public Request authenticate(Route route, Response response)
+            throws IOException
+    {
+        // skip if we already tried or were not asked for Kerberos
+        if (response.request().headers(AUTHORIZATION).stream().anyMatch(SpnegoHandler::isNegotiate) ||
+                response.headers(WWW_AUTHENTICATE).stream().noneMatch(SpnegoHandler::isNegotiate)) {
+            return null;
+        }
+
+        return authenticate(response.request());
+    }
+
+    private static boolean isNegotiate(String value)
+    {
+        return Splitter.on(whitespace()).split(value).iterator().next().equalsIgnoreCase(NEGOTIATE);
+    }
+
+    private Request authenticate(Request request)
+    {
+        byte[] token = generateToken(request.url().host());
+        String credential = format("%s %s", NEGOTIATE, Base64.getEncoder().encodeToString(token));
+        return request.newBuilder()
+                .header(AUTHORIZATION, credential)
+                .build();
+    }
+
+    private byte[] generateToken(String hostName)
+    {
+        String servicePrincipal = makeServicePrincipal(remoteServiceName, hostName, useCanonicalHostname);
+
+        GSSContext context = null;
+        try {
+            Session session = getSession();
+            context = doAs(session.getLoginContext().getSubject(), () -> {
+                GSSContext result = GSS_MANAGER.createContext(
+                        GSS_MANAGER.createName(servicePrincipal, NT_HOSTBASED_SERVICE),
+                        SPNEGO_OID,
+                        session.getClientCredential(),
+                        INDEFINITE_LIFETIME);
+
+                result.requestMutualAuth(true);
+                result.requestConf(true);
+                result.requestInteg(true);
+                result.requestCredDeleg(false);
+                return result;
+            });
+
+            byte[] token = context.initSecContext(new byte[0], 0, 0);
+            if (token != null) {
+                return token;
+            }
+            throw new LoginException("No token generated from GSS context");
+        }
+        catch (GSSException | LoginException e) {
+            throw new ClientException(format("Kerberos error for [%s]: %s", servicePrincipal, e.getMessage()), e);
+        }
+        finally {
+            try {
+                if (context != null) {
+                    context.dispose();
+                }
+            }
+            catch (GSSException ignored) {
+            }
+        }
+    }
+
+    private synchronized Session getSession()
+            throws LoginException, GSSException
+    {
+        if ((clientSession == null) || clientSession.needsRefresh()) {
+            clientSession = createSession();
+        }
+        return clientSession;
+    }
+
+    private Session createSession()
```

Comment:
This is like a pure copy paste from Airlift. I don't think we will ever change HttpClient in Presto. So we need to support both now.